Add support for dispatchJ9Method for MethodHandles on Z

matthewhall2 opened the pull request:
Adds method to OMRNode to determine if node is jitDispatchJ9Method

flushArgs skips first argument for S390HelperCallSnippet call call is jitDipatchJ9Method

needed for https://github.com/eclipse-openj9/openj9/pull/22588. This should not require a coordinated merge since I left the old TR_Debug method in; it will be taken out in another PR.

matthewhall2 commented:
@r30shah can I get a review please?

r30shah commented on compiler/z/codegen/CallSnippet.cpp:
```diff
@@ -72,11 +72,17 @@ uint8_t *TR::S390CallSnippet::S390flushArgumentsToStack(uint8_t *buffer, TR::Nod
     bool rightToLeft = linkage->getRightToLeft() &&
         // we want the arguments for induceOSR to be passed from left to right as in any other non-helper call
-        !callNode->getSymbolReference()->isOSRInductionHelper();
+        !callNode->getSymbolReference()->isOSRInductionHelper() && !isJitDispatchJ9Method;
```
Just a recommendation to increase readability, 
```suggestion
        !(callNode->getSymbolReference()->isOSRInductionHelper() || isJitDispatchJ9Method);
```

r30shah commented on compiler/z/codegen/OMRLinkage.cpp:
```diff
@@ -1623,21 +1623,24 @@ int32_t OMR::Z::Linkage::buildArgs(TR::Node *callNode, TR::RegisterDependencyCon
     bool rightToLeft = self()->isParmsInReverseOrder() &&
         // we want the arguments for induceOSR to be passed from left to right as in any other non-helper call
-        !callNode->getSymbolReference()->isOSRInductionHelper();
+        !callNode->getSymbolReference()->isOSRInductionHelper() &&
```
Similar recommendation as previous.

r30shah reviewed (changes requested):
In general. I feel the changes are fine, so just couple recommendation. 

I have general question about the change though - dispatch through J9method is J9 specific changes, Do we need these in OMR ? Can we make changes in J9 for this ?

r30shah reviewed (approved):
A simple refactoring to facilitate the downstream project to consume some of the code for method dispatch. Changes looks OK to me. I will review the J9 PR next. 
 
@matthewhall2 - Please confirm the order in which this change needs to be delivered.

r30shah commented on compiler/z/codegen/S390HelperCallSnippet.hpp:
```diff
@@ -78,7 +78,13 @@ class S390HelperCallSnippet : public TR::Snippet {
 
     virtual uint8_t *emitSnippetBody();
 
+    uint8_t *emitSnippetBodyInner(uint8_t *cursor, TR::SymbolReference *helperSymRef);
```
Can we use better name for this function ? emitSnippetBodyInner to me seems bit confusing.

r30shah commented on compiler/z/codegen/S390HelperCallSnippet.cpp:
```diff
@@ -162,30 +166,37 @@ uint32_t TR::S390HelperCallSnippet::getLength(int32_t)
-    printSnippetLabel(log, snippet->getSnippetLabel(), bufferPos, "Helper Call Snippet", getName(helperSymRef));
+    TR::SymbolReference *helperSymRef = getHelperSymRef();
+
+    uint8_t *bufferPos = getSnippetLabel()->getCodeLocation();
```
Can you confirm that it prints the same instruction in the log that you are emitting ? Similar to your call `emitSnippetBody` -> `emitSnippetBodyInner` , you have print -> printInner. But I do not see instructions emitted in the emitSnippetBody being printed if it is jitInduceOSR helper.

matthewhall2 replied:
it seems that even before this change the argument flush for `jitInduceOSR` was not printed. I have added it.

r30shah reviewed (commented):
Thanks @matthewhall2  - I have gone through the quick review and recommended some cosmetic changes. I am going to  clone this and OpenJ9 PR and give another look at both PR together. But Overall changes seems fine.

matthewhall2 committed:
Add support for jitDispatchJ9Method for MethodHandles on Z

- breaks up S390HelperCallSnippet::emitSnippetBody into a main and
  helper method; helper to be used by OpenJ9
- breaks up flushArgumentsToStack into a main and a helper; helper to be
  used by OpenJ9

Signed-off-by: Matthew Hall <matthew.hall3@outlook.com>

r30shah reviewed (changes requested):
Overall I changes looks fine to me, some minor nitpick.

r30shah commented on compiler/z/codegen/CallSnippet.cpp:
```diff
@@ -256,6 +268,141 @@ int32_t TR::S390CallSnippet::instructionCountForArguments(TR::Node *callNode, TR
+
+    for (int i = argStart; i < node->getNumChildren(); i++) {
+        TR::Node *child = node->getChild(i);
+        switch (child->getDataType()) {
```
Can this be refactored - I see lot of repetitive code for printing the instruction (Line 303-306, 328-332, 337-341 and later as well? Also please open Issue to added printing of the vector args if not handling it here.

r30shah reviewed (changes requested):
@matthewhall2  - Last nitpick but overall changes looks fine to me

r30shah reviewed (approved):
Thanks @matthewhall2  - Changes looks good to me. One side thought I have about the routine added in this PR for printing instruction in snippet - May be this is something thst can be added in common place like Debug.cpp where we print all instructions. 

But for this PR 👍

matthewhall2 commented:
thanks @r30shah .
> May be this is something that can be added in common place like Debug.cpp where we print all instructions

FYI I believe the Debug APIs are now deprecated (stated at the top of compiler/ras/Debug.hpp). I opened an issue to move some of the other logging to `S390Instruction` (https://github.com/eclipse-omr/omr/issues/8207).

 > Do we need to also check the vector parameter and print the appropriate instruction ? I see the Linkage checks for vector when building args

The current `S390FlushArgumentsToStack` (no changes other than a refactor in this PR) does not flush vector args to the stack, so there is no need to print it.
This doesn't appear to be a problem since we would have noticed it if it were needed, but maybe we should look into it. atm it is used in `OSRInductionHelper`, `generateVIThunk`, `generateInvokeExactJ2IThunk`, and the OpenJ9 side of this PR. I've opened an issue to investigate https://github.com/eclipse-openj9/openj9/issues/23679

matthewhall2 commented:
@0xdaryl can I get a review please?
